File: sdk/typescript/src/risk.ts
/**
 * The risk matrix: likelihood (1–5) x impact (1–5), per OWASP threat, per
 * decision. Port of `keeper_firewall/risk.py` — same bands, same defaults,
 * same residual-vs-inherent rule, so a decision scores identically in both
 * SDKs. The engine only escalates: its verdict joins the decision as one more
 * policy trace and is combined by escalation like everything else.
 */

import { PolicyError } from "./errors.js";
import { THREATS } from "./taxonomy.js";
import type { Action, Finding, PolicyTrace, RiskTier, Stage, ToolCall } from "./types.js";

export type RiskBand = "none" | "low" | "medium" | "high" | "critical";

const BAND_RANK: Record<RiskBand, number> = { none: 0, low: 1, medium: 2, high: 3, critical: 4 };
export const bandRank = (band: RiskBand): number => BAND_RANK[band] ?? 0;

export function bandFor(score: number): RiskBand {
  if (score <= 0) return "none";
  if (score <= 4) return "low";
  if (score <= 9) return "medium";
  if (score <= 16) return "high";
  return "critical";
}

export const DEFAULT_BAND_ACTIONS: Record<RiskBand, Action> = {
  none: "allow",
  low: "allow",
  medium: "flag",
  high: "block",
  critical: "block",
};

export const DEFAULT_TOOL_IMPACT: Record<RiskTier, number> = { low: 2, medium: 3, high: 4, critical: 5 };
export const DEFAULT_LIKELIHOOD_CUTS: [number, number, number, number] = [0.35, 0.5, 0.65, 0.85];

export interface RiskConfig {
  enforce: boolean;
  actions: Record<RiskBand, Action>;
  impactOverrides: Record<string, number>;
  toolImpact: Record<RiskTier, number>;
  likelihoodCuts: [number, number, number, number];
  applicationImpact: Record<string, number>;
}

export const defaultRiskConfig = (): RiskConfig => ({
  enforce: true,
  actions: { ...DEFAULT_BAND_ACTIONS },
  impactOverrides: {},
  toolImpact: { ...DEFAULT_TOOL_IMPACT },
  likelihoodCuts: [...DEFAULT_LIKELIHOOD_CUTS],
  applicationImpact: {},
});

const BANDS: RiskBand[] = ["none", "low", "medium", "high", "critical"];
const ACTIONS: Action[] = ["allow", "flag", "redact", "challenge", "block"];
const TIERS: RiskTier[] = ["low", "medium", "high", "critical"];

function impactValue(value: unknown, where: string): number {
  const n = Number(value);
  if (!Number.isInteger(n) || n < 1 || n > 5) throw new PolicyError(`risk.${where} must be between 1 and 5, got ${String(value)}`);
  return n;
}

/** Parse and validate the `risk:` section of a policy document. */
export function parseRiskConfig(data: unknown): RiskConfig {
  const config = defaultRiskConfig();
  if (!data) return config;
  if (typeof data !== "object") throw new PolicyError("'risk' must be a mapping");
  const doc = data as Record<string, any>;
  const known = new Set(["enforce", "actions", "impact", "tool_impact", "likelihood_cuts", "application_impact"]);
  const unknown = Object.keys(doc).filter((k) => !known.has(k));
  if (unknown.length) throw new PolicyError(`unknown risk keys: ${unknown.sort().join(", ")}`);

  config.enforce = doc.enforce ?? true;
  for (const [band, action] of Object.entries(doc.actions ?? {})) {
    if (!BANDS.includes(band as RiskBand)) throw new PolicyError(`invalid risk section: unknown band '${band}'`);
    if (!ACTIONS.includes(action as Action)) throw new PolicyError(`invalid risk section: unknown action '${String(action)}'`);
    config.actions[band as RiskBand] = action as Action;
  }
  for (const [id, v] of Object.entries(doc.impact ?? {})) {
    if (!(id in THREATS)) throw new PolicyError(`risk.impact: unknown threat id '${id}'`);
    config.impactOverrides[id] = impactValue(v, `impact.${id}`);
  }
  for (const [tier, v] of Object.entries(doc.tool_impact ?? {})) {
    if (!TIERS.includes(tier as RiskTier)) throw new PolicyError(`invalid risk section: unknown tier '${tier}'`);
    config.toolImpact[tier as RiskTier] = impactValue(v, `tool_impact.${tier}`);
  }
  if (doc.likelihood_cuts) {
    const cuts = (doc.likelihood_cuts as unknown[]).map(Number);
    const sorted = [...cuts].sort((a, b) => a - b);
    if (cuts.length !== 4 || cuts.some((c, i) => c !== sorted[i] || !(c > 0 && c <= 1))) {
      throw new PolicyError("risk.likelihood_cuts must be four ascending values in (0, 1]");
    }
    config.likelihoodCuts = cuts as [number, number, number, number];
  }
  for (const [app, v] of Object.entries(doc.application_impact ?? {})) {
    config.applicationImpact[app] = impactValue(v, `application_impact.${app}`);
  }
  return config;
}

export function serialiseRiskConfig(config: RiskConfig): Record<string, unknown> {
  return {
    enforce: config.enforce,
    actions: { ...config.actions },
    impact: { ...config.impactOverrides },
    tool_impact: { ...config.toolImpact },
    likelihood_cuts: [...config.likelihoodCuts],
    application_impact: { ...config.applicationImpact },
  };
}

export interface ThreatRisk {
  threat: string;
  likelihood: number;
  impact: number;
  residualLikelihood: number;
  detectors: string[];
  mitigated: boolean;
}

export interface RiskAssessment {
  likelihood: number;
  impact: number;
  score: number;
  band: RiskBand;
  inherentScore: number;
  inherentBand: RiskBand;
  primary: string | null;
  threats: ThreatRisk[];
  action: Action;
}

export const noRisk = (): RiskAssessment => ({
  likelihood: 0,
  impact: 0,
  score: 0,
  band: "none",
  inherentScore: 0,
  inherentBand: "none",
  primary: null,
  threats: [],
  action: "allow",
});

export function explainRisk(risk: RiskAssessment): string {
  if (!risk.primary) return "no threats identified";
  const title = THREATS[risk.primary]?.title;
  return `risk ${risk.score}/25 (${risk.band}): likelihood ${risk.likelihood} x impact ${risk.impact} on ${risk.primary}${title ? ` ${title}` : ""}`;
}

/** Wire shape, identical to Python's `RiskAssessment.to_dict()`. */
export function serialiseRisk(risk: RiskAssessment): Record<string, unknown> {
  return {
    likelihood: risk.likelihood,
    impact: risk.impact,
    score: risk.score,
    band: risk.band,
    inherent_score: risk.inherentScore,
    inherent_band: risk.inherentBand,
    primary: risk.primary,
    action: risk.action,
    threats: risk.threats.map((c) => ({
      threat: c.threat,
      title: THREATS[c.threat]?.title ?? c.threat,
      framework: THREATS[c.threat]?.framework ?? null,
      likelihood: c.likelihood,
      impact: c.impact,
      score: c.likelihood * c.impact,
      band: bandFor(c.likelihood * c.impact),
      residual_score: c.residualLikelihood * c.impact,
      residual_band: bandFor(c.residualLikelihood * c.impact),
      mitigated: c.mitigated,
      detectors: c.detectors,
    })),
  };
}

export class RiskEngine {
  constructor(readonly config: RiskConfig = defaultRiskConfig()) {}

  likelihood(score: number): number {
    return 1 + this.config.likelihoodCuts.filter((cut) => score >= cut).length;
  }

  impact(threatId: string, _stage: Stage, toolCall?: ToolCall, application?: string): number {
    let base = this.config.impactOverrides[threatId] ?? THREATS[threatId]?.baseImpact ?? 3;
    if (application && application in this.config.applicationImpact) {
      base = Math.max(base, this.config.applicationImpact[application]);
    }
    if (toolCall?.risk) base = Math.max(base, this.config.toolImpact[toolCall.risk] ?? base);
    return Math.max(1, Math.min(5, base));
  }

  assess(findings: Finding[], stage: Stage, options: { toolCall?: ToolCall; application?: string } = {}): RiskAssessment {
    const byThreat = new Map<string, Finding[]>();
    for (const f of findings) {
      if (!f.detected) continue;
      for (const id of f.threats ?? []) {
        if (!byThreat.has(id)) byThreat.set(id, []);
        byThreat.get(id)!.push(f);
      }
    }
    const cells: ThreatRisk[] = [];
    for (const [threat, fs] of byThreat) {
      let likelihood = this.likelihood(Math.max(...fs.map((f) => f.score)));
      const detectors = [...new Set(fs.map((f) => f.detector))];
      if (detectors.length >= 2) likelihood = Math.min(5, likelihood + 1);
      const mitigated = fs.every((f) => f.action === "redact");
      cells.push({
        threat,
        likelihood,
        impact: this.impact(threat, stage, options.toolCall, options.application),
        residualLikelihood: mitigated ? 1 : likelihood,
        detectors,
        mitigated,
      });
    }
    if (!cells.length) return noRisk();

    // Stable sort: on ties the first-mapped threat stays the headline.
    const ordered = cells
      .map((c, i) => ({ c, i }))
      .sort(
        (a, b) =>
          b.c.residualLikelihood * b.c.impact - a.c.residualLikelihood * a.c.impact ||
          b.c.likelihood * b.c.impact - a.c.likelihood * a.c.impact ||
          a.i - b.i,
      )
      .map((x) => x.c);
    const worst = ordered[0];
    const inherent = ordered.reduce((m, c) => (c.likelihood * c.impact > m.likelihood * m.impact ? c : m), ordered[0]);
    const score = worst.residualLikelihood * worst.impact;
    const band = bandFor(score);
    return {
      likelihood: worst.residualLikelihood,
      impact: worst.impact,
      score,
      band,
      inherentScore: inherent.likelihood * inherent.impact,
      inherentBand: bandFor(inherent.likelihood * inherent.impact),
      primary: worst.threat,
      threats: ordered,
      action: this.config.actions[band] ?? "allow",
    };
  }

  trace(risk: RiskAssessment, policyId: string, policyVersion: string, dryRun = false): PolicyTrace | null {
    if (!this.config.enforce || risk.action === "allow") return null;
    return {
      policyId,
      policyVersion,
      ruleId: `risk-matrix/${risk.band}`,
      matched: true,
      action: dryRun ? "allow" : risk.action,
      elapsedMs: 0,
      note: (dryRun ? `[dry-run: would ${risk.action}] ` : "") + explainRisk(risk),
    };
  }
}

/** Count (likelihood, impact) pairs into a 5x5 grid, `grid[likelihood-1][impact-1]`. */
export function riskMatrix(cells: [number, number][]): number[][] {
  const grid = Array.from({ length: 5 }, () => Array(5).fill(0) as number[]);
  for (const [l, i] of cells) if (l >= 1 && l <= 5 && i >= 1 && i <= 5) grid[l - 1][i - 1] += 1;
  return grid;
}

File: sdk/python/keeper_firewall/risk.py
"""The risk matrix: likelihood x impact, per OWASP threat, per decision.

Detectors answer "did I see an attack?". Security teams need a different
answer: "how bad is this, *here*?". A weak injection signal in a chat about the
weather and the same weak signal driving a ``payment.transfer`` tool call are
the same finding and very different risks. The risk engine turns findings into
a classic 5x5 risk matrix cell so that:

* enforcement can escalate on *context* (a medium-confidence signal on a
  critical sink is blocked, the same signal on a read-only path is flagged);
* every audit event carries one comparable number (1-25) and band, which is
  what the dashboard heat map, alerting and SIEM severity are built on;
* the answer is explainable — "L3 x I5 on ASI02 Tool Misuse" — not a model's
  opinion.

**Likelihood (1-5)** is how confident the evidence is that the threat is real,
taken from the strongest finding mapped to the threat. Boundary trust is
already inside detector scores (the injection detector multiplies by it), so it
is not counted twice here. Two or more *independent* detectors agreeing on a
threat adds one step: corroboration is real evidence.

**Impact (1-5)** is how bad it is if the threat succeeds: the threat's base
impact (policy-overridable), raised to the sink's risk tier when a tool call is
involved, and to the application's criticality floor when one is configured.
Persistence is already priced into the threats that describe it (ASI06), so a
memory write is not bumped a second time.

**Residual vs inherent.** A finding whose action is ``redact`` has already been
mitigated: the sensitive span will not reach the model. The *inherent* risk is
recorded for reporting, and the *residual* risk — likelihood reset to 1 for
fully mitigated threats — is what drives enforcement. Without that, redacting a
phone number would score as a critical LLM02 event and block the request the
redaction was meant to let through.

The engine only ever *escalates*. Its verdict joins the decision as one more
policy trace and is combined by escalation like everything else, so the risk
matrix can turn a flag into a block but can never turn a detector's block into
an allow.
"""

from __future__ import annotations

import enum
from collections.abc import Iterable, Mapping, Sequence
from dataclasses import dataclass, field
from typing import Any

from .errors import PolicyError
from .taxonomy import THREATS
from .types import Action, Finding, PolicyTrace, RiskTier, Stage, ToolCall


class RiskBand(str, enum.Enum):
    NONE = "none"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

    def rank(self) -> int:
        return _BAND_RANK[self]

    @classmethod
    def for_score(cls, score: int) -> RiskBand:
        if score <= 0:
            return cls.NONE
        if score <= 4:
            return cls.LOW
        if score <= 9:
            return cls.MEDIUM
        if score <= 16:
            return cls.HIGH
        return cls.CRITICAL


_BAND_RANK = {RiskBand.NONE: 0, RiskBand.LOW: 1, RiskBand.MEDIUM: 2, RiskBand.HIGH: 3, RiskBand.CRITICAL: 4}

DEFAULT_BAND_ACTIONS: Mapping[RiskBand, Action] = {
    RiskBand.NONE: Action.ALLOW,
    RiskBand.LOW: Action.ALLOW,
    RiskBand.MEDIUM: Action.FLAG,
    RiskBand.HIGH: Action.BLOCK,
    RiskBand.CRITICAL: Action.BLOCK,
}

DEFAULT_TOOL_IMPACT: Mapping[RiskTier, int] = {
    RiskTier.LOW: 2,
    RiskTier.MEDIUM: 3,
    RiskTier.HIGH: 4,
    RiskTier.CRITICAL: 5,
}

#: Finding score cut-points for likelihood 2, 3, 4 and 5.
DEFAULT_LIKELIHOOD_CUTS: tuple[float, float, float, float] = (0.35, 0.5, 0.65, 0.85)


@dataclass(slots=True)
class RiskConfig:
    """The ``risk:`` section of a policy document."""

    enforce: bool = True
    actions: Mapping[RiskBand, Action] = field(default_factory=lambda: dict(DEFAULT_BAND_ACTIONS))
    impact_overrides: Mapping[str, int] = field(default_factory=dict)
    tool_impact: Mapping[RiskTier, int] = field(default_factory=lambda: dict(DEFAULT_TOOL_IMPACT))
    likelihood_cuts: tuple[float, float, float, float] = DEFAULT_LIKELIHOOD_CUTS
    #: Per-application impact floor, e.g. ``{"payments-agent": 4}``: an asset's
    #: criticality raises the impact of every threat against it.
    application_impact: Mapping[str, int] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, data: Mapping[str, Any] | None) -> RiskConfig:
        if not data:
            return cls()
        if not isinstance(data, Mapping):
            raise PolicyError("'risk' must be a mapping")
        unknown = set(data) - {"enforce", "actions", "impact", "tool_impact", "likelihood_cuts", "application_impact"}
        if unknown:
            raise PolicyError(f"unknown risk keys: {', '.join(sorted(unknown))}")
        try:
            actions = dict(DEFAULT_BAND_ACTIONS)
            for band, action in (data.get("actions") or {}).items():
                actions[RiskBand(str(band))] = Action(str(action))
            impact = {str(k): _impact(v, f"impact.{k}") for k, v in (data.get("impact") or {}).items()}
            for tid in impact:
                if tid not in THREATS:
                    raise PolicyError(f"risk.impact: unknown threat id {tid!r}")
            tool_impact = dict(DEFAULT_TOOL_IMPACT)
            for tier, v in (data.get("tool_impact") or {}).items():
                tool_impact[RiskTier(str(tier))] = _impact(v, f"tool_impact.{tier}")
            raw_cuts = [float(x) for x in (data.get("likelihood_cuts") or DEFAULT_LIKELIHOOD_CUTS)]
            if len(raw_cuts) != 4 or raw_cuts != sorted(raw_cuts) or not all(0.0 < c <= 1.0 for c in raw_cuts):
                raise PolicyError("risk.likelihood_cuts must be four ascending values in (0, 1]")
            cuts = (raw_cuts[0], raw_cuts[1], raw_cuts[2], raw_cuts[3])
            app_impact = {str(k): _impact(v, f"application_impact.{k}")
                          for k, v in (data.get("application_impact") or {}).items()}
        except ValueError as exc:
            raise PolicyError(f"invalid risk section: {exc}") from exc
        return cls(
            enforce=bool(data.get("enforce", True)),
            actions=actions,
            impact_overrides=impact,
            tool_impact=tool_impact,
            likelihood_cuts=cuts,
            application_impact=app_impact,
        )


def _impact(value: Any, where: str) -> int:
    n = int(value)
    if not 1 <= n <= 5:
        raise PolicyError(f"risk.{where} must be between 1 and 5, got {n}")
    return n


@dataclass(slots=True)
class ThreatRisk:
    """The matrix cell for one threat within one decision."""

    threat: str
    likelihood: int
    impact: int
    residual_likelihood: int
    detectors: tuple[str, ...]
    mitigated: bool = False

    @property
    def score(self) -> int:
        return self.likelihood * self.impact

    @property
    def residual_score(self) -> int:
        return self.residual_likelihood * self.impact

    @property
    def band(self) -> RiskBand:
        return RiskBand.for_score(self.score)

    @property
    def residual_band(self) -> RiskBand:
        return RiskBand.for_score(self.residual_score)

    def to_dict(self) -> dict[str, Any]:
        t = THREATS.get(self.threat)
        return {
            "threat": self.threat,
            "title": t.title if t else self.threat,
            "framework": t.framework if t else None,
            "likelihood": self.likelihood,
            "impact": self.impact,
            "score": self.score,
            "band": self.band.value,
            "residual_score": self.residual_score,
            "residual_band": self.residual_band.value,
            "mitigated": self.mitigated,
            "detectors": list(self.detectors),
        }


@dataclass(slots=True)
class RiskAssessment:
    """The decision-level risk: the worst residual threat cell.

    ``score``/``band`` are residual (what enforcement acted on);
    ``inherent_score``/``inherent_band`` are before mitigation.
    """

    likelihood: int = 0
    impact: int = 0
    score: int = 0
    band: RiskBand = RiskBand.NONE
    inherent_score: int = 0
    inherent_band: RiskBand = RiskBand.NONE
    primary: str | None = None
    threats: tuple[ThreatRisk, ...] = ()
    action: Action = Action.ALLOW

    @property
    def cell(self) -> tuple[int, int]:
        return (self.likelihood, self.impact)

    def to_dict(self) -> dict[str, Any]:
        return {
            "likelihood": self.likelihood,
            "impact": self.impact,
            "score": self.score,
            "band": self.band.value,
            "inherent_score": self.inherent_score,
            "inherent_band": self.inherent_band.value,
            "primary": self.primary,
            "action": self.action.value,
            "threats": [t.to_dict() for t in self.threats],
        }

    def explain(self) -> str:
        if not self.primary:
            return "no threats identified"
        t = THREATS.get(self.primary)
        title = f" {t.title}" if t else ""
        return (
            f"risk {self.score}/25 ({self.band.value}): likelihood {self.likelihood} x impact "
            f"{self.impact} on {self.primary}{title}"
        )


class RiskEngine:
    """Scores fired findings onto the matrix and proposes an action."""

    def __init__(self, config: RiskConfig | None = None) -> None:
        self.config = config or RiskConfig()

    def likelihood(self, score: float) -> int:
        level = 1
        for cut in self.config.likelihood_cuts:
            if score >= cut:
                level += 1
        return level

    def impact(
        self,
        threat_id: str,
        stage: Stage,
        *,
        tool_call: ToolCall | None = None,
        application: str | None = None,
    ) -> int:
        threat = THREATS.get(threat_id)
        base = self.config.impact_overrides.get(threat_id, threat.base_impact if threat else 3)
        if application and application in self.config.application_impact:
            base = max(base, self.config.application_impact[application])
        if tool_call is not None and tool_call.risk is not None:
            base = max(base, self.config.tool_impact.get(tool_call.risk, base))
        return max(1, min(5, base))

    def assess(
        self,
        findings: Iterable[Finding],
        stage: Stage,
        *,
        tool_call: ToolCall | None = None,
        application: str | None = None,
    ) -> RiskAssessment:
        by_threat: dict[str, list[Finding]] = {}
        for f in findings:
            if not f.detected:
                continue
            for tid in f.threats:
                by_threat.setdefault(tid, []).append(f)

        cells: list[ThreatRisk] = []
        for tid, fs in by_threat.items():
            likelihood = self.likelihood(max(f.score for f in fs))
            detectors = tuple(dict.fromkeys(f.detector for f in fs))
            if len(detectors) >= 2:
                likelihood = min(5, likelihood + 1)
            mitigated = all(f.action is Action.REDACT for f in fs)
            cells.append(
                ThreatRisk(
                    threat=tid,
                    likelihood=likelihood,
                    impact=self.impact(tid, stage, tool_call=tool_call, application=application),
                    residual_likelihood=1 if mitigated else likelihood,
                    detectors=detectors,
                    mitigated=mitigated,
                )
            )
        if not cells:
            return RiskAssessment()

        # Stable sort: on ties the first-mapped threat (the detector's primary
        # category, e.g. LLM01 before ASI01) stays the headline.
        cells.sort(key=lambda c: (-c.residual_score, -c.score))
        worst = cells[0]
        inherent = max(cells, key=lambda c: c.score)
        band = worst.residual_band
        return RiskAssessment(
            likelihood=worst.residual_likelihood,
            impact=worst.impact,
            score=worst.residual_score,
            band=band,
            inherent_score=inherent.score,
            inherent_band=inherent.band,
            primary=worst.threat,
            threats=tuple(cells),
            action=self.config.actions.get(band, Action.ALLOW),
        )

    def trace(self, assessment: RiskAssessment, *, policy_id: str, policy_version: str,
              dry_run: bool = False) -> PolicyTrace | None:
        """The assessment as a policy trace, or ``None`` if it changes nothing."""
        if not self.config.enforce or assessment.action is Action.ALLOW:
            return None
        return PolicyTrace(
            policy_id=policy_id,
            policy_version=policy_version,
            rule_id=f"risk-matrix/{assessment.band.value}",
            matched=True,
            action=Action.ALLOW if dry_run else assessment.action,
            note=(f"[dry-run: would {assessment.action.value}] " if dry_run else "") + assessment.explain(),
        )


def matrix(cells: Sequence[tuple[int, int]]) -> list[list[int]]:
    """Count (likelihood, impact) pairs into a 5x5 grid, ``grid[likelihood-1][impact-1]``."""
    grid = [[0] * 5 for _ in range(5)]
    for likelihood, impact in cells:
        if 1 <= likelihood <= 5 and 1 <= impact <= 5:
            grid[likelihood - 1][impact - 1] += 1
    return grid


__all__ = [
    "DEFAULT_BAND_ACTIONS",
    "RiskAssessment",
    "RiskBand",
    "RiskConfig",
    "RiskEngine",
    "ThreatRisk",
    "matrix",
]
